File: e2e/load-person.mjs
// Bounded authenticated HTTP load, reusing the isolated operational E2E seed.
import assert from 'node:assert/strict';
import { readFileSync, writeFileSync } from 'node:fs';
import { request } from '@playwright/test';
import { Client } from 'pg';
import { captureSecrets, remember } from './support/evidence.mjs';

const contexts = [];
const db = new Client({ connectionString: process.env.E2E_AUDIT_URL });
const seed = JSON.parse(readFileSync('/private/seed.json'));
const pool = JSON.parse(readFileSync('/private/load-pool.json'));
const password = process.env.E2E_PASSWORD;
const agents = [];
const durationSeconds = Number(process.env.E2E_LOAD_DURATION_SECONDS || 30);
const agentCount = Number(process.env.E2E_LOAD_AGENTS || 100);
const poolSize = Number(process.env.E2E_LOAD_POOL_SIZE || 25000);
assert(Number.isInteger(durationSeconds) && durationSeconds >= 2 && durationSeconds <= 120);
assert(Number.isInteger(agentCount) && agentCount >= 1 && agentCount <= 100);
assert.equal(pool.total_people, poolSize);
assert.equal(pool.person_ids.length, pool.target_pool_size);
async function context() {
  const ctx = await request.newContext({ baseURL: 'http://web:8080', timeout: 10000 });
  contexts.push(ctx); return ctx;
}
async function post(ctx, path, data, status) {
  const response = await ctx.post(path, { data });
  await captureSecrets(response);
  assert.equal(response.status(), status, path);
  const body = await response.json();
  if (body.accept_path) remember(body.accept_path.split('/').at(-1));
  await response.dispose(); return body;
}
async function snapshot() {
  const { rows } = await db.query(`SELECT
    (SELECT coalesce(sum(calls),0)::float8 FROM pg_stat_statements WHERE userid=(SELECT oid FROM pg_roles WHERE rolname='crm_app')) AS executions,
    (SELECT coalesce(sum(total_exec_time),0)::float8 FROM pg_stat_statements WHERE userid=(SELECT oid FROM pg_roles WHERE rolname='crm_app')) AS execution_ms,
    xact_commit::float8,xact_rollback::float8,blks_read::float8,blks_hit::float8,temp_bytes::float8,deadlocks::float8
    FROM pg_stat_database WHERE datname=current_database()`);
  return rows[0];
}
async function activity() {
  const { rows } = await db.query(`SELECT count(*)::int AS connections,
    count(*) FILTER (WHERE state='active')::int AS active,
    count(*) FILTER (WHERE state='idle in transaction')::int AS idle_in_transaction,
    count(*) FILTER (WHERE wait_event_type='Lock')::int AS lock_waiters
    FROM pg_stat_activity WHERE usename='crm_app'`);
  return { at_ms: Date.now(), ...rows[0] };
}
function percentile(values, p) { return values[Math.max(0, Math.ceil(values.length * p) - 1)] ?? null; }
function randomPersonIndex(state) {
  // Independent deterministic streams: reproducible random traversal without
  // serial cycles or one global RNG shared by all agents.
  state.value = (Math.imul(state.value, 1664525) + 1013904223) >>> 0;
  return state.value % pool.person_ids.length;
}
async function phase(name, selected, durationMs) {
  const before = await snapshot(), samples = [], requests = [];
  const started = Date.now(), deadline = performance.now() + durationMs;
  let monitoring = true;
  const monitor = (async () => {
    while (monitoring) { samples.push(await activity()); await new Promise(r => setTimeout(r, 500)); }
  })();
  await Promise.all(selected.map(async (agent, index) => {
    const state = { value: (0x9e3779b9 ^ ((index + 1) * 0x85ebca6b)) >>> 0 };
    while (performance.now() < deadline) {
      const personIndex = randomPersonIndex(state);
      const person = pool.person_ids[personIndex];
      const start = performance.now();
      let status = 0, valid = false, payloadBytes = 0;
      try {
        const response = await agent.ctx.get('/api/people/' + person);
        status = response.status();
        payloadBytes = Number(response.headers()['content-length'] || 0);
        if (status === 200) {
          const body = await response.json();
          valid = body.person?.id === person
            && body.contact_methods?.length === pool.expected_detail.contact_methods
            && body.tags?.length === pool.expected_detail.tags
            && body.tasks?.length === pool.expected_detail.tasks
            && body.history?.length === pool.expected_detail.history;
        }
        await response.dispose();
      } catch { /* Count transport/timeouts without logging request credentials. */ }
      requests.push({ agent: index + 1, person_index: personIndex + 1, status, valid,
        payload_bytes: payloadBytes, elapsed_ms: performance.now() - start });
    }
  }));
  const finished = Date.now();
  monitoring = false; await monitor;
  const after = await snapshot();
  const timings = requests.map(r => r.elapsed_ms).sort((a,b) => a-b);
  const payloads = requests.filter(r => r.payload_bytes > 0).map(r => r.payload_bytes).sort((a,b) => a-b);
  const result = { name, agents: selected.length, started_ms: started, finished_ms: finished,
    seconds: (finished-started)/1000, requests: requests.length,
    successful: requests.filter(r => r.status === 200 && r.valid).length,
    invalid_200: requests.filter(r => r.status === 200 && !r.valid).length,
    statuses: Object.fromEntries([...new Set(requests.map(r => r.status))].map(status => [status, requests.filter(r => r.status === status).length])),
    requests_per_second: requests.length/((finished-started)/1000),
    latency_ms: { p50: percentile(timings,.5), p95: percentile(timings,.95), p99: percentile(timings,.99), max: timings.at(-1) },
    payload_bytes: { p50: percentile(payloads,.5), max: payloads.at(-1) ?? null },
    postgres_delta: Object.fromEntries(Object.keys(before).map(key => [key, after[key]-before[key]])),
    activity: samples, samples: requests };
  writeFileSync('/artifacts/load-' + name + '.json', JSON.stringify(result, null, 2));
  console.log(JSON.stringify({ phase: name, agents: selected.length, requests: result.requests,
    successful: result.successful, statuses: result.statuses, rps: result.requests_per_second, latency_ms: result.latency_ms }));
  return result;
}

try {
  await db.connect();
  const admin = await context();
  await post(admin, '/api/session', { email: seed.actors.admin.email, password }, 200);
  for (let index=0; index<agentCount; index++) {
    const ctx = await context();
    const email = 'load-agent-' + index + '@e2e.test';
    const invitation = await post(admin, '/api/organization/invitations', { email, role: 'member' }, 201);
    const identity = await post(ctx, '/api/invitations/accept', {
      token: invitation.accept_path.split('/').at(-1), display_name: 'Load Agent ' + index, password }, 200);
    assert.equal(identity.organization.id, seed.actors.admin.organization.id);
    agents.push({ ctx, user: identity.user.id });
  }
  assert.equal(new Set(agents.map(agent => agent.user)).size, agentCount);
  const foreign = await context();
  await post(foreign, '/api/session', { email: seed.actors.casey.email, password }, 200);
  const denied = await foreign.get('/api/people/' + pool.person_ids[0]);
  assert.equal(denied.status(), 404); await denied.dispose();
  for (const agent of agents) {
    const response = await agent.ctx.get('/api/people/' + pool.person_ids[0]);
    assert.equal(response.status(),200);
    const body = await response.json();
    assert.equal(body.history.length, pool.expected_detail.history);
    assert.equal(body.tasks.length, pool.expected_detail.tasks);
    await response.dispose();
  }
  const results = [];
  results.push(await phase('baseline', agents.slice(0,1), durationSeconds * 1000));
  results.push(await phase('hundred-agents', agents, durationSeconds * 1000));
  for (const result of results) {
    result.person_coverage = [];
    for (let index = 1; index <= result.agents; index++) {
      const visits = result.samples.filter(sample => sample.agent === index);
      const distinct = new Set(visits.map(sample => sample.person_index)).size;
      result.person_coverage.push({ agent: index, distinct_persons: distinct, requests: visits.length });
    }
  }
  const finalState = (await db.query(`SELECT (SELECT count(*)::int FROM person) AS people,
    (SELECT count(*)::int FROM inquiry) AS inquiries,
    (SELECT count(*)::int FROM inquiry_received) AS inquiry_facts,
    (SELECT count(*)::int FROM stage_changed) AS stage_facts,
    (SELECT count(*)::int FROM note) AS notes,
    (SELECT count(*)::int FROM task) AS tasks,
    (SELECT count(*)::int FROM contact_attempted) AS contact_facts,
    (SELECT count(*)::int FROM person_detail_projection) AS projections`)).rows[0];
  pool.expected_state.projections = pool.target_pool_size;
  assert.deepEqual(finalState, pool.expected_state);
  writeFileSync('/artifacts/load-summary.json', JSON.stringify({ fixture: `${agentCount} distinct member identities, one Organization, ${poolSize} Persons, ${pool.target_pool_size} ${pool.fixture_kind} random-access targets`,
    workload: 'Each agent uses an independent deterministic pseudo-random stream over the configured random-access target pool on every request. Closed-loop HTTP GET through nginx and the real API; one in-flight request per agent; no think time; no browser rendering or UI fanout',
    duration_seconds_per_condition: durationSeconds, pool_size: poolSize, agent_count: agentCount,
    sql_profile: false, foreign_tenant_denied: true, final_state: finalState,
    phases: results.map(({ samples, activity, ...rest }) => rest) }, null, 2));
} finally { await db.end(); for (const ctx of contexts) await ctx.dispose(); }

File: e2e/load_person.py
"""Run a bounded large-pool Person-detail measurement in an owned E2E stack.

Usage: python3 -m e2e.load_person
"""
import argparse
import json
import re
import secrets
import shutil
import signal
import threading
import time
import uuid
from e2e import runner


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--duration-seconds', type=int, default=30,
                        help='duration of each 1-agent and 100-agent condition (default 30)')
    parser.add_argument('--agents', type=int, default=100)
    parser.add_argument('--pool-size', type=int, default=25000)
    parser.add_argument('--database-connections', type=int, default=10,
                        help='API SQLx maximum connection count (default 10)')
    parser.add_argument('--fixture', choices=['data-light', 'history-heavy'], default='data-light')
    parser.add_argument('--heavy-pool-size', type=int, default=1000,
                        help='random-access hot set when --fixture history-heavy (default 1000)')
    options = parser.parse_args()
    if not 2 <= options.duration_seconds <= 120:
        parser.error('--duration-seconds must be between 2 and 120')
    if not 1 <= options.agents <= 100:
        parser.error('--agents must be between 1 and 100')
    if not 1 <= options.pool_size <= 25000:
        parser.error('--pool-size must be between 1 and 25000')
    if not 1 <= options.database_connections <= 200:
        parser.error('--database-connections must be between 1 and 200')
    if options.fixture == 'history-heavy' and not 1 <= options.heavy_pool_size <= options.pool_size:
        parser.error('--heavy-pool-size must be between 1 and --pool-size')
    run_id = uuid.uuid4().hex[:12]
    project = f'crm-e2e-{run_id}-personload-1-a1'
    artifacts = runner.STATE / 'runs' / run_id / 'personload-1-a1'
    private = runner.STATE / 'private' / project
    artifacts.mkdir(parents=True, mode=0o700)
    private.mkdir(parents=True, mode=0o700)
    for sig in (signal.SIGINT, signal.SIGTERM):
        signal.signal(sig, lambda *_: runner.STOP.set())
    credentials = {key: secrets.token_hex(32) for key in ['app','root','migrator','audit','session','raw','password',
        'realtime_api','realtime_token','inbound','inference','livekit_key','livekit_secret']}
    images = runner.build(artifacts.parent, targets=['api','web','browser'])
    manifest = runner.config(project, private, artifacts, images, credentials, 'leads')
    manifest['services']['postgres']['command'] = ['postgres', '-c', 'shared_preload_libraries=pg_stat_statements']
    manifest['services']['api']['environment']['CRM_DATABASE_MAX_CONNECTIONS'] = str(options.database_connections)
    load = manifest['services']['browser']
    load['command'] = ['node', 'load-person.mjs']
    load['environment'].update({'E2E_LOAD_DURATION_SECONDS': str(options.duration_seconds),
                                'E2E_LOAD_AGENTS': str(options.agents),
                                'E2E_LOAD_POOL_SIZE': str(options.pool_size)})
    load['volumes'].append(f"{runner.ROOT / 'e2e/load-person.mjs'}:/e2e/load-person.mjs:ro")
    # Deliberately leave SQLx per-query profiling off during the latency test.
    runner.write_json(private / 'compose.json', manifest)
    runner.write_json(private / 'centrifugo.json', {'log': {'level':'info'}, 'health': {'enabled':True},
        'client': {'allowed_origins':['http://web:8080']}, 'channel': {'namespaces':[{'name':'org'}]}})
    for path in private.iterdir(): path.chmod(0o600)
    shutil.copyfile(runner.ROOT / 'e2e/load-person.mjs', artifacts / 'workload.mjs')
    cmd = runner.compose(project, private)
    def run(args, timeout=180, log='runner.log'):
        runner.command(cmd + args, artifacts / log, timeout=timeout)
    report = {'project':project, 'started_at':time.time(), 'status':'failed',
              'revision':runner.read_command(['git','rev-parse','HEAD']), 'images':images,
              'database_max_connections': options.database_connections,
              'scripts': {name: runner.hashlib.sha256((runner.ROOT/name).read_bytes()).hexdigest()
                          for name in ['e2e/load_person.py', 'e2e/load-person.mjs']},
              'docker_resources': runner.read_command(['docker','info','--format',
                  '{{.NCPU}} CPUs; {{.MemTotal}} bytes RAM; {{.Architecture}}; Docker {{.ServerVersion}}'])}
    monitor_stop = threading.Event()
    monitor = None
    try:
        run(['up','-d','--wait','postgres','centrifugo','mocks'])
        run(['exec','-T','postgres','psql','-X','-v','ON_ERROR_STOP=1','-U','postgres','-d','crm','-c',
             'CREATE EXTENSION pg_stat_statements; GRANT pg_read_all_stats TO e2e_audit;'])
        run(['run','--rm','migrate'])
        run(['run','--rm','bootstrap'])
        run(['up','-d','--wait','api','web'])
        run(['run','--rm','seed'])
        # Only these disposable read-pool rows bypass commands. Creating 25k
        # rows through ReceiveInquiry would benchmark its serial intake lock,
        # instead of the requested Person-detail read path.
        pool_row = runner.read_command(cmd + ['exec','-T','postgres','psql','-X','-A','-t','-v','ON_ERROR_STOP=1',
            '-U','postgres','-d','crm','-c',
            "SELECT o.id || '|' || (SELECT id FROM stage WHERE organization_id=o.id AND position=1) || '|' || (SELECT id FROM stage WHERE organization_id=o.id AND position=2) || '|' || (SELECT user_id FROM organization_membership WHERE organization_id=o.id AND role='admin' ORDER BY user_id LIMIT 1) FROM organization o WHERE o.name='Journey Realty'"])
        org_id, stage_id, second_stage_id, admin_id = pool_row.split('|')
        if not all(re.fullmatch(r'[0-9a-f-]{36}', value) for value in [org_id, stage_id, second_stage_id, admin_id]):
            raise RuntimeError('large-pool fixture could not resolve organization/stage')
        pool_sql = "INSERT INTO person (organization_id,first_name,last_name,stage_id) SELECT '" + org_id + "'::uuid,'Load','Pool ' || n::text,'" + stage_id + "'::uuid FROM generate_series(1," + str(options.pool_size) + ") AS n"
        run(['exec','-T','postgres','psql','-X','-v','ON_ERROR_STOP=1',
             '-U','postgres','-d','crm','-c',pool_sql], timeout=120, log='pool-seed.log')
        expected_detail = {'contact_methods': 0, 'tags': 0, 'tasks': 0, 'history': 0}
        expected_state = {'people': options.pool_size, 'inquiries': 0, 'inquiry_facts': 0,
                          'stage_facts': 0, 'notes': 0, 'tasks': 0, 'contact_facts': 0}
        target_limit = options.pool_size
        if options.fixture == 'history-heavy':
            target_limit = options.heavy_pool_size
            heavy_sql = f"""
WITH heavy AS (SELECT id,row_number() OVER (ORDER BY id) AS n FROM person WHERE organization_id='{org_id}'::uuid ORDER BY id LIMIT {target_limit})
INSERT INTO contact_method(organization_id,person_id,kind,value,normalized_value)
SELECT '{org_id}'::uuid,h.id,CASE WHEN g=1 THEN 'email' ELSE 'phone' END,
       CASE WHEN g=1 THEN 'heavy-'||h.n||'@example.test' ELSE '+1202'||lpad((h.n*10+g)::text,7,'0') END,
       CASE WHEN g=1 THEN 'heavy-'||h.n||'@example.test' ELSE '+1202'||lpad((h.n*10+g)::text,7,'0') END
FROM heavy h CROSS JOIN generate_series(1,3) g;
WITH heavy AS (SELECT id FROM person WHERE organization_id='{org_id}'::uuid ORDER BY id LIMIT {target_limit})
INSERT INTO note(organization_id,person_id,author_user_id,body,origin,correlation_id,created_at,updated_at)
SELECT '{org_id}'::uuid,h.id,'{admin_id}'::uuid,'Production-shaped relationship note '||g||' with enough content to represent an ordinary agent interaction and follow-up context.','web_session',gen_random_uuid(),now()-(g||' days')::interval,now()-(g||' days')::interval
FROM heavy h CROSS JOIN generate_series(1,12) g;
WITH heavy AS (SELECT id FROM person WHERE organization_id='{org_id}'::uuid ORDER BY id LIMIT {target_limit})
INSERT INTO task(organization_id,person_id,title,kind,due_at,assignee_user_id,created_by_user_id,completed_at,completed_by_user_id,origin,correlation_id,created_at,updated_at)
SELECT '{org_id}'::uuid,h.id,'Production-shaped task '||g,CASE g%5 WHEN 0 THEN 'call' WHEN 1 THEN 'email' WHEN 2 THEN 'text' WHEN 3 THEN 'follow_up' ELSE 'other' END,now()+(g||' days')::interval,'{admin_id}'::uuid,'{admin_id}'::uuid,CASE WHEN g>12 THEN now()-(g||' hours')::interval END,CASE WHEN g>12 THEN '{admin_id}'::uuid END,'web_session',gen_random_uuid(),now()-(g||' days')::interval,now()-(g||' days')::interval
FROM heavy h CROSS JOIN generate_series(1,20) g;
WITH heavy AS (SELECT id FROM person WHERE organization_id='{org_id}'::uuid ORDER BY id LIMIT {target_limit})
INSERT INTO contact_attempted(organization_id,actor_kind,actor_user_id,on_behalf_of_user_id,origin,occurred_at,recorded_at,correlation_id,person_id,channel,outcome)
SELECT '{org_id}'::uuid,'user','{admin_id}'::uuid,'{admin_id}'::uuid,'web_session',now()-(g||' days')::interval,now()-(g||' days')::interval,gen_random_uuid(),h.id,CASE g%4 WHEN 0 THEN 'call' WHEN 1 THEN 'text' WHEN 2 THEN 'email' ELSE 'other' END,CASE g%4 WHEN 0 THEN 'reached' WHEN 1 THEN 'sent' WHEN 2 THEN 'left_message' ELSE 'no_answer' END
FROM heavy h CROSS JOIN generate_series(1,20) g;
WITH heavy AS (SELECT id FROM person WHERE organization_id='{org_id}'::uuid ORDER BY id LIMIT {target_limit})
INSERT INTO stage_changed(organization_id,actor_kind,actor_user_id,on_behalf_of_user_id,origin,occurred_at,recorded_at,correlation_id,person_id,from_stage_id,to_stage_id,reason)
SELECT '{org_id}'::uuid,'user','{admin_id}'::uuid,'{admin_id}'::uuid,'web_session',now()-((g+20)||' days')::interval,now()-((g+20)||' days')::interval,gen_random_uuid(),h.id,CASE WHEN g%2=0 THEN '{stage_id}'::uuid ELSE '{second_stage_id}'::uuid END,CASE WHEN g%2=0 THEN '{second_stage_id}'::uuid ELSE '{stage_id}'::uuid END,'load_fixture'
FROM heavy h CROSS JOIN generate_series(1,10) g;
WITH heavy AS (SELECT id FROM person WHERE organization_id='{org_id}'::uuid ORDER BY id LIMIT {target_limit})
INSERT INTO assignment_changed(organization_id,actor_kind,actor_user_id,on_behalf_of_user_id,origin,occurred_at,recorded_at,correlation_id,person_id,from_user_id,to_user_id,reason)
SELECT '{org_id}'::uuid,'user','{admin_id}'::uuid,'{admin_id}'::uuid,'web_session',now()-((g+30)||' days')::interval,now()-((g+30)||' days')::interval,gen_random_uuid(),h.id,NULL,'{admin_id}'::uuid,'load_fixture'
FROM heavy h CROSS JOIN generate_series(1,10) g;
INSERT INTO tag(organization_id,name,created_by_user_id) SELECT '{org_id}'::uuid,'Heavy Tag '||g,'{admin_id}'::uuid FROM generate_series(1,5) g;
WITH heavy AS (SELECT id FROM person WHERE organization_id='{org_id}'::uuid ORDER BY id LIMIT {target_limit})
INSERT INTO person_tag(organization_id,person_id,tag_id,added_by_user_id) SELECT '{org_id}'::uuid,h.id,t.id,'{admin_id}'::uuid FROM heavy h CROSS JOIN tag t WHERE t.organization_id='{org_id}'::uuid AND t.name LIKE 'Heavy Tag %';
"""
            run(['exec','-T','postgres','psql','-X','-v','ON_ERROR_STOP=1','-U','postgres','-d','crm','-c',heavy_sql], timeout=180, log='heavy-seed.log')
            expected_detail = {'contact_methods': 3, 'tags': 5, 'tasks': 12, 'history': 60}
            expected_state.update({'stage_facts': target_limit * 10, 'notes': target_limit * 12,
                                   'tasks': target_limit * 20, 'contact_facts': target_limit * 20})
        rebuild_sql = f"SELECT crm_rebuild_person_detail_projection('{org_id}'::uuid,id) FROM person WHERE organization_id='{org_id}'::uuid ORDER BY id LIMIT {target_limit}"
        run(['exec','-T','postgres','psql','-X','-v','ON_ERROR_STOP=1',
             '-U','postgres','-d','crm','-c',rebuild_sql], timeout=180, log='projection-seed.log')
        person_ids = runner.read_command(cmd + ['exec','-T','postgres','psql','-X','-A','-t','-v','ON_ERROR_STOP=1',
            '-U','postgres','-d','crm','-c',f"SELECT id FROM person WHERE organization_id='{org_id}'::uuid ORDER BY id LIMIT {target_limit}"]).splitlines()
        if len(person_ids) != target_limit or any(len(value) != 36 for value in person_ids):
            raise RuntimeError('large-pool fixture count or IDs did not match request')
        runner.write_json(private / 'load-pool.json', {'organization_id': org_id, 'person_ids': person_ids,
            'total_people': options.pool_size, 'target_pool_size': target_limit,
            'fixture_kind': options.fixture, 'expected_detail': expected_detail,
            'expected_state': expected_state})
        ids = runner.read_command(cmd + ['ps','-q','api','postgres']).split()
        def sample():
            with (artifacts / 'container-stats.jsonl').open('w') as stream:
                while not monitor_stop.is_set():
                    try:
                        stats = runner.read_command(['docker','stats','--no-stream','--format','{{json .}}',*ids])
                        for line in stats.splitlines():
                            stream.write(json.dumps({'at_ms':int(time.time()*1000), **json.loads(line)})+'\n')
                        stream.flush()
                    except Exception as error:
                        stream.write(json.dumps({'error':type(error).__name__})+'\n'); stream.flush()
                    monitor_stop.wait(.5)
        monitor = threading.Thread(target=sample, daemon=True); monitor.start()
        run(['run','--rm','browser'], timeout=150, log='load.log')
        report['status'] = 'completed'
    finally:
        monitor_stop.set()
        if monitor: monitor.join(timeout=35)
        report['finished_at'] = time.time()
        try:
            runner.command(cmd + ['logs','--no-color','--timestamps'], artifacts/'services.log', timeout=30, cancellable=False)
        finally:
            runner.owned_cleanup(project, private, artifacts/'cleanup.log')
            report['cleanup'] = 'verified_empty'
            runner.sanitize(artifacts, private, credentials)
            shutil.rmtree(private)
            runner.write_json(artifacts/'run.json',report)
        print(f"Load evidence: {artifacts}", flush=True)


if __name__ == '__main__':
    main()
